Nested volumes and explicit <url> tags

mjpost opened the pull request:
Started consolidating URLs per #156. The goal is to:

- [x] Convert all `href` attributes to `<url>` tags
- [x] Convert all `<href>` tags to `<url>` tags
- [x] Use relative IDs for Anthology identifies (e.g., `P18-1014` instead of `https://www.aclweb.org/anthology/P18-1014`)

Adding explicit `<url>` tags is easier with nested volumes (#317), so I am doing that, too:

- [x] Convert schemas to support nesting
- [x] Update all the XML
- [x] Add explicit full-volume `<url>` tags for volumes that are present on the server
- [x] Update the site generation code to work with the nested volumes

Additionally I think we should:

- [x] remove `<bibtype>` and `<bibkey>` tags, which are just clutter
- [x] Reformat all the XML to be properly indented (2 spaces? 2 spaces)

This will help with mirroring (#295 #28 #22).

Finally, I'd like to update the code not to generate URLs for papers without a URL tag, so by cross-checking against #264, we can solve a lot of problems (#226 #181 #180).

mjpost commented:
Another issue has come up. We now need to generate explicit volume `<url>` tags. This will be easier once #317 is done, so I will move to that next.

mjpost commented:
Update: this is taking quite a bit longer than I expected, but I am making progress, and believe I am nearly there. If you want to look at the code I changed in `bin/anthology/*.py` to ensure that I haven't done terrible damage to the codebase, your time wouldn't be wasted at this point. I can very nearly generate the exact YAML from the nested format (which XML files are not yet checked in) as from the flat format with the code in the `master` branch.

mbollmann commented on bin/anthology/volumes.py:
```diff
@@ -63,46 +86,44 @@ def _set_meta_info(self):
                 self.attrib["meta_date"] = "{}/{}".format(self.get("year"), month)
-        if is_journal(self.top_level_id):
+        if is_journal(self.collection_id):
+            print('** ATTRIB', self.attrib)
```
You can always use `log.debug()` for these things, then you don't need to remember to take them out later ;)

mbollmann commented on bin/anthology/anthology.py:
```diff
@@ -54,38 +54,35 @@ def import_directory(self, importdir):
+                log.critical("Triggered by file: {}".format(filename))
+
+            # front matter
+            if len(volume.content):
```
This reads a bit obscure, I needed some time to figure out how/why it does the right thing. Maybe `volume` can keep an attribute `has_frontmatter` or something?

mbollmann replied:
I still feel this should be changed to be more semantic :)

mjpost replied:
Good idea, done.

mbollmann commented on bin/anthology/anthology.py:
```diff
@@ -54,38 +54,35 @@ def import_directory(self, importdir):
             self.import_file(xmlfile)
         self.pindex.verify()
-        
+
```
This reminds me of something I've did rigorously but never made explicit: I used [black](https://black.readthedocs.io/en/stable/) to keep the formatting style consistent. Do you think it would be a good idea to add this to the README as "coding guidelines" or something?

mjpost replied:
I do think this is a good idea and we could implement it.

mbollmann commented on bin/anthology/papers.py:
```diff
@@ -187,18 +136,26 @@ def _interpret_pages(self):
+        else:
+            # If not, only the last three
+            _id = "{}{:03d}".format(self.volume_id, int(self._id))
+        return _id
```
When IDs can now have a variable number of digits, let's add an `assert len(_id) == 4` or something just to be sure.

mjpost replied:
Good idea.

akoehn commented:
Can you do a squash-merge to keep the changes more local? I am a big fan of rewriting history (in git at least ...) so that it is easier to see where a feature was introduced.

That is just my personal opinion of course.

mjpost commented:
Yes, will do a squash merge. This is a good idea.

@acl-org/anthology, this is ready for review. Here's a summary of changes:

This commit:

- Introduces a nested format (closes #317)
- Only displays URLs if they are found in the XML
- URLs are stored using a relative format for internal links (closes #156)
- Thereby facilitates mirroring (#295)
- Volume PDFs are displayed only if present in the <meta> block.
- Enables front matter for journals that have URLs (closes #181).
- Added missing entries and removed ones whose PDFs were missing, including LREC 2014 (closes #31)
- Add links to journal frontmatter where they exist (addresses #264)
- I inserted URLs via a scripted check for volume PDFs
- Punts on C69 formatting (closes #147)

Enabled, but yet to do:
- Creating PDF volumes from individual papers (#226)

mbollmann commented:
I want to build this locally and have a closer look at it later today.

akoehn commented:
@mbollmann: To have a look at the results (rather than the source), build master, `mv build build_master`, build this branch and compare with meld (or another diff tool).  Tried this for other PRs and that workflow works quite well.

akoehn commented:
There is now a benoit_sagot and benoit_sagot1:

> title: "Beno\u0131\u0302t Sagot"
> title: "Beno\xEEt Sagot"

Maybe it was there before and now fixed in master but not yet in this PR?

mjpost commented:
@akoehn this kind of diffing is what I did. I did bulk diffs on `.md` and `.bib` files, and found only differences of this sort (which I think will be resolved when we do the final merge).

I have merged in `master` a few times but I think there may be one that's missing.

davidweichiang commented:
Good catch. I think this was fixed in master, though. The issue is that this uses a dotless i and Unicode normalization doesn't know how to deal with that. I don't think there's anywhere in the code that fixes this automatically, so I'll create a new issue for this.

mbollmann commented on hugo/layouts/events/single.html:
```diff
@@ -30,7 +30,7 @@ <h4 class="card-title">Contents</h4>
     <h4 class="pb-2">
       <span class="mr-2">
-      {{- with $volume.url -}}
+      {{- if $volume.url -}}
```
That looks wrong. What happened here?

mbollmann replied:
But that's exactly what the `with...` statement does. It is functionally the same as `if isset ...`, except that it sets the dot `.` to the variable content within the block as well.

In your current version, the code is simply wrong, as the `.` on the next line won't be set to `$volume.url` anymore.

mbollmann commented on hugo/layouts/papers/list-entry.html:
```diff
@@ -14,9 +14,11 @@
     </a>
     {{- end -}}
   {{- else -}}
+    {{- if isset $paper "url" -}}
```
Should be `with $paper.url` and a dot in place of `$paper.url` in the next line

mjpost replied:
Would your proposed change handle if there were no `url` member?

mbollmann replied:
Yes.

mbollmann commented on hugo/layouts/papers/single.html:
```diff
@@ -17,7 +17,9 @@
   {{ with $volume.meta_date }}<meta content="{{ . }}" name="citation_publication_date" >{{ end }}
 {{ end }}
-<meta content="{{ $paper.url }}" name="citation_pdf_url" >
+{{ if $paper.url }}
```
Same as above, use `with ...`

mbollmann commented on hugo/layouts/volumes/single.html:
```diff
@@ -92,7 +92,7 @@ <h2 id="title">
     <!-- Most of the styling for this block is set in _papers.scss to avoid clutter -->
     <div class="acl-paper-link-block">
-      {{ with $paper.url }}
+      {{ if $paper.url }}
```
Again, this seems wrong.

mbollmann commented on bin/anthology/volumes.py:
```diff
@@ -118,7 +161,7 @@ def get_title(self, form="xml"):
           - html:  Convert XML tags into valid HTML tags
         """
-        return self.formatter(self.get("xml_title"), form)
+        return self.formatter(self.get("xml_booktitle"), form)
```
Can you explain this change? Maybe it's obvious but I don't quite follow.

mjpost replied:
The title of a volume is `booktitle`, stored in the metadata. I think the volume title used to be read in as the `title` field from paper ID 0.

mbollmann replied:
Ah, right, this is coming from the metadata block now. Got it, I think.

mbollmann commented on hugo/layouts/papers/single.html:
```diff
@@ -173,9 +175,11 @@ <h5 class="card-title">Abstract</h5>
         </a>
         {{ end }}
       {{ else }}
+        {{ if isset $paper "url" }}
```
This could still be changed to `with`, but that's more of a style thing than a functional one :)

mjpost replied:
Whoops, missed that one. Just pushed. Thank you!

akoehn commented:
Congratulations on this huge merge!
